### PR TITLE
feat(product-exporter): new package

### LINKED ISSUE #291: Implement a product exporter module
### Description

Implement a module to retrieve commercetools products from the CTP

(Part of a series implementing the rewrite of the [product-csv-sync](https://github.com/sphereio/sphere-node-product-csv-sync) module)

- [x] Expose a public method to receive the output stream
- [x] Recursively get products 
- [x] Write products to the output stream
- [x] Implement a CLI to get input from users
- [x] Add documentation
- [x] Add integration tests

### PR DESCRIPTION
**Pull request by wizzy25**
#### Summary

Add new package to export products in JSON format from the CTP

resolves #291 

#### Todo

- [x] Expose a public method to receive the output stream
- [x] Recursively get products 
- [x] Write products to the output stream
- [x] Implement a CLI to get input from users
- [x] Add documentation
- [x] Add integration tests
- [x] Increase test coverage
- [x] Properly define product type
- [x] Switch to another logger

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,6 +8,9 @@ node_modules
 .npm
 .released-packages
 
+# Integration tests dependencies cache
```
Comment (hisabimbola) [inline .gitignore]: can we just merge this to master in another small PR or commit, else every commit to master will have to add this

File: docs/cli/product-exporter.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
+# Product Exporter
+
+A package that helps with exporting [commercetools products](https://dev.commercetools.com/http-api-projects-products.html) from the [commercetools platform](http://dev.commercetools.com/).
+The products can be exported in JSON format, or as chunks that can be piped to a parser for more export formats.
```
[the remark this answers is not in the document]
Comment (wizzy25) [reply docs/cli/product-exporter.md]: Well, yes. 
`or as chunks that can be piped to a parser for more export formats`

File: docs/cli/product-exporter.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+## Configuration
+
+The constructor accepts four arguments:
+- `apiConfig` (Object): `AuthMiddleware` options for authentication on the commercetools platform. (Required. See [here](https://commercetools.github.io/nodejs/sdk/api/sdkMiddlewareAuth.html#named-arguments-options))
+- `exportConfig` (Object): Internal Export configurations
+  - `batch` (Number): Amount of products to fetch for each API call
+  - `expand` (String): Reference fields to expand in the returned product
+  - `json` (Boolean): Specify if products returned should be in JSON file format. If set to false, the products will be output in chunks (Default: true)
```
[the remark this answers is not in the document]
Comment (wizzy25) [reply docs/cli/product-exporter.md]: Any idea how to make it easier to understand? A tutorial maybe...

File: packages/product-exporter/package.json [hunk 4]
```diff
@@ -0,0 +1,48 @@
+{
+  "name": "@commercetools/product-exporter",
+  "version": "1.0.0-development",
+  "description": "Export products from the commercetools platform",
+  "keywords": [
+    "commercetools",
+    "products",
+    "projections",
+    "parser",
```
[the remark this answers is not in the document]
Comment (wizzy25) [reply packages/product-exporter/package.json]: Oops 😄

File: packages/product-exporter/src/cli.js [hunk 5]
```diff
@@ -0,0 +1,194 @@
+import fs from 'fs'
+import { getCredentials } from '@commercetools/get-credentials'
+import npmlog from 'npmlog'
```
[the remark this answers is not in the document]
Comment (wizzy25) [reply packages/product-exporter/src/cli.js]: Concerning this, I'd like us to discuss about it first

File: integration-tests/cli/product-export.it.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+      productType.id = createdProductType[0].body.id
+      taxCategory.id = createdTaxCategory[0].body.id
+
+      const sampleProducts = createProducts(
+        productType,
+        taxCategory,
+      )
+      return createData(apiConfig, 'products', sampleProducts)
+    })
+    .catch(process.stderr.write)
```
Comment (hisabimbola) [inline integration-tests/cli/product-export.it.js]: why are you _catching_ here, I think if an error occurs it's not going to exit out of the test

File: packages/product-exporter/src/cli.js [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+  // Limit batchSize to 500 in accord with the API
+  .coerce('batchSize', (arg) => {
+    const batchSize = parseInt(arg, 10)
+    if (batchSize <= 0 || batchSize > 500)
+      throw new Error('Invalid batchSize, must be a number between 1 and 500')
+
+    return batchSize
+  })
+
+  .option('expand', {
```
[the remark this answers is not in the document]
Comment (wizzy25) [reply packages/product-exporter/src/cli.js]: I'd add a detailed description in the documentation. If that's not enough, we can think about making a tutorial to make things as easy as possible

File: integration-tests/cli/product-export.it.js [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+      apiConfig = {
+        host: 'https://auth.sphere.io',
+        apiUrl: 'https://api.sphere.io',
+        projectKey,
+        credentials,
+      }
+      return clearData(apiConfig, 'products')
+    })
+    .then(() => (
+      Promise.all([
```
[the remark this answers is not in the document]
Comment (wizzy25) [reply integration-tests/cli/product-export.it.js]: Well I'm not already using async here, but Yay!

File: .gitignore (hunk 1 above)
Comment (wizzy25) [reply .gitignore]: Merged in #304

File: docs/cli/product-exporter.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+```
+Usage: product-exporter [options]
+Export products from the commercetools platform
+
+Options:
+  --help, -h        Show help text.                                    [boolean]
+  --version, -v     Show version number                                [boolean]
+  --projectKey, -p  API project key                          [string] [required]
+  --apiUrl          The host URL of the HTTP API service
+                                     [string] [default: "https://api.sphere.io"]
```
Comment (hisabimbola) [inline docs/cli/product-exporter.md]: change this url to `https://api.commercetools.com`, the `sphere.io` one has issues.

File: docs/cli/product-exporter.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+  --authUrl         The host URL of the OAuth API service
+                                    [string] [default: "https://auth.sphere.io"]
```
Comment (hisabimbola) [inline docs/cli/product-exporter.md]: this also should be `https://auth.commercetools.com`

File: docs/cli/product-exporter.md [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+  - If passed `false` (or omitted), only published products are retrieved
+
+### JS
+For more direct usage, it is possible to use this module directly:
+```js
+import ProductExporter from '@commercetools/product-exporter'
+import fs from 'fs'
+
+const apiConfig = {
+  host: 'https://api.sphere.io',
```
Comment (hisabimbola) [inline docs/cli/product-exporter.md]: the host and apiUrl here too should change as specified above

File: packages/product-exporter/package.json [hunk 12]
```diff
@@ -0,0 +1,47 @@
+{
+  "name": "@commercetools/product-exporter",
+  "version": "1.0.0-development",
+  "description": "Export products from the commercetools platform",
+  "keywords": [
+    "commercetools",
+    "products",
+    "projections",
+    "price",
```
Comment (hisabimbola) [inline packages/product-exporter/package.json]: is the "price" keyword relevant here?

File: packages/product-exporter/src/main.js [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+  run (outputStream: stream$Writable): Promise<*> {
+    this.logger.debug('Starting Export')
+    const formattedStream = ProductExporter._decideStream(
+      this.exportConfig.exportType,
+    )
+    this.logger.debug('Preparing outputStream')
+    formattedStream.pipe(outputStream)
+    return this._getProducts(formattedStream)
+    .catch((e: Error) => {
+      this.logger.debug(e, 'Oops. Something went wrong')
```
Comment (hisabimbola) [inline packages/product-exporter/src/main.js]: I think the log level for this should be `error` since we are logging error. Any reason why you used `debug`?

File: packages/product-exporter/src/main.js [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+      method: 'GET',
+    }
+    if (this.accessToken)
+      request.headers = {
+        Authorization: `Bearer ${this.accessToken}`,
+      }
+
+    const processConfig: Object = { accumulate: false }
+    if (this.exportConfig.total)
+      processConfig.total = this.exportConfig.total
```
Comment (hisabimbola) [inline packages/product-exporter/src/main.js]: we can log this to the logger(debug)

File: packages/product-exporter/src/main.js [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+
+    this.logger.debug('Dispatching request')
+    return this.client.process(request, (
+      {
+        body: {
+          results: products,
+        },
+      }: ProcessFnResponse,
+    ): Promise<*> => {
+      this.logger.debug(`Fetched ${products.length} products`)
```
Comment (hisabimbola) [inline packages/product-exporter/src/main.js]: 👍

File: packages/product-exporter/src/main.js [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+      projectKey,
+    }).productProjections
+    service.staged(exportConfig.staged)
+
+    if (exportConfig.batch)
+      service.perPage(exportConfig.batch)
+    if (exportConfig.predicate)
+      service.where(exportConfig.predicate)
+    // Handle `expand` separately because it's an array
+    if (exportConfig.expand)
```
Comment (hisabimbola) [inline packages/product-exporter/src/main.js]: you might want to check length here instead. `exportConfig.expand.length`

File: packages/product-exporter/src/main.js [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+  }
+
+  /* if the exportFormat is json, prepare the stream for json data. If
+  csv, also create a json stream because it needs to pass text to
+  the stdout, but the json format preparation is irrelevant this time
+  */
+  static _decideStream (exportType: 'json' | 'chunk') {
+    return exportType === 'json'
+      ? JSONStream.stringify('[\n', ',\n', '\n]')
+      : JSONStream.stringify('', '\n\n\n', '\n\n')
```
Comment (hisabimbola) [inline packages/product-exporter/src/main.js]: I think it might be beneficial to add comment as to why we are doing this

File: packages/product-exporter/src/main.js (hunk 14 above)
Comment (wizzy25) [reply packages/product-exporter/src/main.js]: I don't get what exactly should be logged here

File: docs/cli/product-exporter.md (hunk 9 above)
Comment (wizzy25) [reply docs/cli/product-exporter.md]: I get this error with `http://api.commercetools.io`:

`ERR: request to https://auth.commercetools.io/oauth/token failed, reason: getaddrinfo ENOTFOUND auth.commercetools.io auth.commercetools.io:443]`

Comment (hisabimbola) [reply docs/cli/product-exporter.md]: it's `api.commercetools.com` not `io`

File: integration-tests/cli/helpers/products-export.data.js [hunk 18]
```diff
@@ -0,0 +1,194 @@
+export const sampleProductType = {
+  name: 'productTypeForProductExport',
+  description: 'bla bla bla',
+  key: `productTypeKey${Date.now()}`,
+}
+
+export const sampleTaxCategory = {
+  name: 'new-sample-tax-category',
+  key: `tax-cat-1${Date.now()}`,
```
Comment (adnasa) [inline integration-tests/cli/helpers/products-export.data.js]: it looks like this file is used for testing purposes.
I would use a hard coded value here instead of `Date.now()`

File: integration-tests/cli/product-export.it.js [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+  createProducts,
+  expectedProducts,
+} from './helpers/products-export.data'
+import { createData, clearData } from './helpers/utils'
+
+let projectKey
+if (process.env.CI === 'true')
+  projectKey = 'product-export-integration-test'
+else
+  projectKey = process.env.npm_config_projectkey
```
Comment (adnasa) [inline integration-tests/cli/product-export.it.js]: wouldn't this be `npm_config_projectKey`?

File: packages/product-exporter/src/cli.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (adnasa) [reply packages/product-exporter/src/cli.js]: a link to the documentation would suffice for me.
http://dev.commercetools.com/http-api.html#reference-expansion

also a comment here would add value

File: packages/product-exporter/src/main.js [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+      })
+
+    return service.build()
+  }
+
+  /* if the exportFormat is json, prepare the stream for json data. If
+  csv, also create a json stream because it needs to pass text to
+  the stdout.
+  */
+  static _decideStream (exportType: 'json' | 'chunk') {
```
Comment (adnasa) [inline packages/product-exporter/src/main.js]: why this weird naming?
Isn't this simply a getter?

suggestion: `_getStream`

File: packages/product-exporter/test/main.spec.js [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+    const exportConfig = {
+      staged: true,
+      batch: 5,
+      predicate: 'foo=bar',
+      expand: ['someReference', 'anotherReference'],
+    }
+
+    it('should build uri with query options', () => {
+      const expectedUri = oneLineTrim`
+        /my-project-key/product-projections?staged=true&expand=someReference
```
Comment (adnasa) [inline packages/product-exporter/test/main.spec.js]: nitpick if you are using oneLinetrim, you might as well separate the slashes into an new line:

```js
oneLineTrim`
/all
/the
/values
/trimmed
/into
/one
/line
`
```

Comment (hisabimbola) [reply packages/product-exporter/test/main.spec.js]: We still need to make it readable though not one word per line.

File: integration-tests/cli/product-export.it.js (hunk 19 above)
Comment (hisabimbola) [reply integration-tests/cli/product-export.it.js]: I think npm lowercase it's arguments.

File: integration-tests/cli/helpers/products-export.data.js (hunk 18 above)
Comment (hisabimbola) [reply integration-tests/cli/helpers/products-export.data.js]: I think this is to ensure that the key is unique. Can we bypass, but I think this is okay. No hard opinion.

Comment (wizzy25) [reply integration-tests/cli/helpers/products-export.data.js]: This is to ensure the key is unique. In edge cases, not doing so could cause the tests to fail

File: integration-tests/cli/product-export.it.js (hunk 19 above)
Comment (wizzy25) [reply integration-tests/cli/product-export.it.js]: It's all lower case :-)

**Comment by wizzy25** [comment]
Let's get this merged :1st_place_medal:
